### PR TITLE
Fix stuck ACs

### PR DESCRIPTION
**Pull request by Atlante45**
This PR attempts to solve an issue where will stay stuck after the sandbox shuts down and prevent users from upgrading until they manually kill those processes.
It creates a Windows Process Group for the ACs when they are spawned by a monitor.
Windows will kill all the processes in that group when the monitor dies.

### CODE REVIEW

File: libraries/shared/src/SharedUtil.cpp [hunk 1]
```diff
@@ -1081,7 +1081,7 @@ void setMaxCores(uint8_t maxCores) {
 void quitWithParentProcess() {
     if (qApp) {
         qDebug() << "Parent process died, quitting";
-        qApp->quit();
+        exit(0);
```
Comment (Atlante45) [inline libraries/shared/src/SharedUtil.cpp]: I've kept the old system in place for the other OS and have also upgraded it to a direct shutdown.

**Comment by Atlante45** [comment]
Although this should fix the issue, I think we should revisit this at some point.
We could use that job object to have all our processes appear in the same collapsible group in the Task manager.
The metadata of our binaries is also lacking and we should fill it out.

I think fixing those 2 things would leave users less confused when looking for our services in their task manager, or trying to understand what all those "assignment-client.exe" processes are and why they need them.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2885/HighFidelity-Beta-PR11269-71cf1ab40efb4f4369788f131a2a86700c07b8ee.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2892/HighFidelity-Beta-PR11269-db7fce2f6404502d38cedd81d1f8cf48c1f3e618.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2892/HighFidelity-Beta-PR11269-db7fce2f6404502d38cedd81d1f8cf48c1f3e618.exe

**Comment by Atlante45** [comment]
# Test plan:
Restart/Shutdown the sandbox a dozen times and verify there are no stuck ACs.

**Comment by ShoaibSmith** [comment]
Testing

**Comment by ShoaibSmith** [comment]
Pass. I am not seeing any Assignment Clients running after I Restart/Shutdown Sandbox.

File: libraries/shared/src/SharedUtil.cpp [hunk 2]
```diff
@@ -1113,3 +1113,56 @@ void watchParentProcess(int parentPID) {
     timer->start();
 }
 #endif
+
+
+#ifdef Q_OS_WIN
+QString GetLastErrorAsString() {
+    //Get the error message, if any.
+    DWORD errorMessageID = ::GetLastError();
+    if (errorMessageID == 0)
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.cpp]: Style: missing curly braces

File: libraries/shared/src/SharedUtil.cpp [hunk 3]
```diff
@@ -1113,3 +1113,56 @@ void watchParentProcess(int parentPID) {
...
+        return QString(); //No error message has been recorded
+
+    LPSTR messageBuffer = nullptr;
+    size_t size = FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL, errorMessageID, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPSTR)&messageBuffer, 0, NULL);
+
+    std::string message(messageBuffer, size);
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.cpp]: Can we create the QString here?

File: libraries/shared/src/SharedUtil.cpp [hunk 4]
```diff
@@ -1113,3 +1113,56 @@ void watchParentProcess(int parentPID) {
     timer->start();
 }
 #endif
+
+
+#ifdef Q_OS_WIN
+QString GetLastErrorAsString() {
+    //Get the error message, if any.
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.cpp]: Based on the format I'm guessing these comments were copied, but I'd prefer we just delete them - they add no value IMO.

File: libraries/shared/src/SharedUtil.cpp [hunk 5]
```diff
@@ -1113,3 +1113,56 @@ void watchParentProcess(int parentPID) {
...
+
+    std::string message(messageBuffer, size);
+
+    //Free the buffer.
+    LocalFree(messageBuffer);
+
+    return QString::fromStdString(message);
+}
+
+void *createJobObject() {
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.cpp]: Thoughts on changing the return type to `HANDLE`? If not, the `*`  should be moved to the left.

File: libraries/shared/src/SharedUtil.h [hunk 6]
```diff
@@ -238,4 +238,10 @@ void setMaxCores(uint8_t maxCores);
 const QString PARENT_PID_OPTION = "parent-pid";
 void watchParentProcess(int parentPID);
 
+
+#ifdef _WIN32
+void *createJobObject();
+void addProcessToJobObject(void *jobObject, qint64 processId);
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.h]: Should the `void *` here be `HANDLE`?

File: libraries/shared/src/SharedUtil.h [hunk 7]
```diff
@@ -238,4 +238,10 @@ void setMaxCores(uint8_t maxCores);
 const QString PARENT_PID_OPTION = "parent-pid";
 void watchParentProcess(int parentPID);
 
+
+#ifdef _WIN32
```
Comment (huffman) [inline libraries/shared/src/SharedUtil.h]: For consistency, I think we should use `Q_OS_WIN` here

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2952/HighFidelity-Beta-PR11269-557cf8c3960d2af7166b89f13f4eac732741216c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2952/HighFidelity-Beta-PR11269-557cf8c3960d2af7166b89f13f4eac732741216c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2956/HighFidelity-Beta-PR11269-26b4e22340fd4f7177d026309dab989ba0e206dd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2956/HighFidelity-Beta-PR11269-26b4e22340fd4f7177d026309dab989ba0e206dd.exe

**Comment by huffman** [comment]
Closing for now, we will revisit it later.

**Pull request closed without merging**